src/lib.rs

```diff
@@ -1299,16 +1299,73 @@ impl Build {
         // must be done in parallel so that acquire won't block forever.
         let wait_thread = thread::Builder::new().spawn(move || {
-            for (cmd, program, mut child, _token) in rx {
-                wait_on_child(&cmd, &program, &mut child.0)?;
-            }
+            let mut error = None;
+            let mut pendings = Vec::new();
+            // Buffer the stdout
+            let mut stdout = io::BufWriter::with_capacity(128, io::stdout());
+
+            loop {
+                let mut has_made_progress = false;
+
+                // Reading new pending tasks
+                loop {
+                    match rx.try_recv() {
+                        Ok(pending) => {
+                            has_made_progress = true;
+                            pendings.push(pending)
+                        }
+                        Err(mpsc::TryRecvError::Disconnected) if pendings.is_empty() => {
+                            let _ = stdout.flush();
+                            return if let Some(err) = error {
+                                Err(err)
+                            } else {
+                                Ok(())
+                            };
+                        }
+                        _ => break,
+                    }
+                }
+
+                // Try waiting on them.
+                pendings.retain_mut(|(cmd, program, child, _)| {
+                    match try_wait_on_child(cmd, program, &mut child.0, &mut stdout) {
+                        Ok(Some(())) => {
+                            // Task done, remove the entry
+                            has_made_progress = true;
+                            false
+                        }
+                        Ok(None) => true, // Task still not finished, keep the entry
+                        Err(err) => {
+                            // Task fail, remove the entry.
+                            has_made_progress = true;
 
-            Ok(())
+                            // Since we can only return one error, log the error to make
+                            // sure users always see all the compilation failures.
+                            let _ = writeln!(stdout, "cargo:warning={}", err);
+                            error = Some(err);
+
+                            false
+                        }
+                    }
+                });
+
+                if !has_made_progress {
+                    // Given that we spawned a lot of compilation tasks, it is unlikely
+                    // that OS cannot find other ready task to execute.
+                    //
+                    // If all of them are done, then we will yield them and spawn more,
+                    // or simply returns.
+                    //
+                    // Thus this will not be turned into a busy-wait loop and it will not
+                    // waste CPU resource.
+                    thread::yield_now();
```

Comment:
Hmm, I really don't like this. We're waiting on another process though and not another thread, so I'm not sure there's a much better approach.

That said, I would like this to keep a backoff count, and start sleeping for a bit if progress hasn't been made rather than yielding.

Reply:
Thanks, I've added a backoff count and sleep for `[100, 1000]` milliseconds if no progress at all for at least three loops.

> We're waiting on another process though and not another thread, so I'm not sure there's a much better approach.

The only ergonomic solution I can think of is using asynchronous instead of synchronous semantics, that will avoid all the annoying deadlock caused by blocking operations here and also remove needs for threading at all.

That's why I now think having a small async runtime in std is really a good idea, I used to think that's terrible since one in libstd probably won't match performance from tokio and pulling tokio should be straight-forward and ok with tokio, but now I realize that it's not always possible to pull dependencies and having battery in libstd does help in this case.